modify scatter_update to hack variable initialization check

suiyuan2009 opened the pull request:
Fixes #3373

suiyuan2009 committed:
modify scatter_update to hack variable initialization check

tensorflow-jenkins commented:
Can one of the admins verify this patch?

Mistobaan commented:
no tests? Also, it looks very hacky in the way it forces initialization.

suiyuan2009 commented:
I'll add test. I have tested on cpu, and it truly reduced memory cost. I find that if use `sess.run(data.initializer)`, memory cost doubles because init_variable will stay in memory.

suiyuan2009 committed:
add test and doc

keveman commented:
@suiyuan2009 Thanks a lot for the PR and sorry for the late response. There is lot of good work gone into this CL. However, a kernel initializing an input ref tensor has no precedent elsewhere in TensorFlow. Also, adding @alextp to comment as he might have more insight into this.

alextp commented:
@suiyuan2009 Thanks for writing this!

It sounds like the problem you're trying to fix here is that the memory usage for zero-initializing variables is bigger than what it takes to store the variable. Is that it?

If so, I wonder if there's a better way of doing it than hacking scatter_add.

suiyuan2009 commented:
@alextp modifying tf.Variable is a choice.

suiyuan2009 committed:
Revert "add test and doc"

This reverts commit be3796f23d79e8189debe4418da87a3713fdff03.

suiyuan2009 committed:
Revert "modify scatter_update to hack variable initialization check"

This reverts commit 20fb524c94425a7726dcf88635ce382e7b6a1d12.

alextp commented:
@suiyuan2009 another option is to make a ZeroInitialize op which allocates storage and zero-initializes it (the current assign op will allocate storage but it needs the zero-tensor to exist).

@mrry @keveman  do you think it makes sense to support this?

mrry commented:
@alextp I think your suggestion of a ZeroInitializer op makes more sense than adding a flag to the individual assign ops. I'd be inclined to keep it in `tf.contrib` because it seems specific to the current implementation of variables, it's not fundamental, and we will probably want to deprecate it once there's no longer a need for it.

Longer term, it would be great if we could identify points in the graph where a tensor buffer is expiring, so we could replace the copy in `tf.assign()` with a move, but in the short term this should work.

alextp commented:
@suiyuan2009 do you want to implement this ZeroInitialize op? I expect it
will be quite similar to what you have here in the end.

On Wed, Sep 7, 2016 at 9:42 AM, Derek Murray <notifications@github.com>
wrote:

> @alextp <https://github.com/alextp> I think your suggestion of a
> ZeroInitializer op makes more sense than adding a flag to the individual
> assign ops. I'd be inclined to keep it in tf.contrib because it seems
> specific to the current implementation of variables, it's not fundamental,
> and we will probably want to deprecate it once there's no longer a need for
> it.
>
> Longer term, it would be great if we could identify points in the graph
> where a tensor buffer is expiring, so we could replace the copy in
> tf.assign() with a move, but in the short term this should work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4077#issuecomment-245342032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxUGcDQo6ifSM9-uT_CYwDn4eCo3dks5qnulcgaJpZM4JuwXP>
> .
>



-- 
 - Alex

suiyuan2009 commented:
@alextp I'll implement it.

suiyuan2009 committed:
add ZeroInitializer Op

suiyuan2009 commented:
@alextp ZeroInitializerOp is added.

alextp commented on tensorflow/contrib/framework/python/ops/variables.py:
```diff
@@ -50,9 +51,29 @@
+           'zero_initializer']
 
 
+def zero_initializer(ref, use_locking=True, name="zero_initializer"):
```
Is this wrapper necessary? I think we should be able to just forward the generated op.

suiyuan2009 replied:
I want to put this op in tf.contrib...

alextp commented on tensorflow/contrib/framework/python/ops/variables_test.py:
```diff
@@ -810,5 +810,27 @@ def testInitFromCheckpointWithScopes(self):
+    var0 = tf.Variable(tf.zeros(shape, dtype=dtype))
+    var1 = tf.Variable(tf.ones(shape, dtype=dtype))
+    var0_zero = tf.contrib.framework.zero_initializer(var0)
+    var1_zero = tf.contrib.framework.zero_initializer(var1)
```
Why an explicit zero initializer for a variable defined around tf.ones? Why not just var0 in this test?

alextp replied:
Indeed, maybe we should make this a RepeatedValueInitializer, with the value specified as an attr, so it can be used to replace both tf.zeros and tf.ones as initializer.

suiyuan2009 replied:
just to check that all has been initialized with zero. I'll make initializer as a param. What is a RepeatedValueInitializer?

alextp replied:
My suggestion was that you make an op which takes any constant float value (0.0 by default) and initializes the variable to have that value repeated in all entries. Then you can replace also things which are not zero initialized (like your ones-initialized variable, which weirdly gets initialized with zero by your zero_initializer op).

I'd also like it if the syntax was more like var0 = tf.Variable(tf.zero_initializer(shape, dtype)) to make it work seamlessly with the existing infrastructure such as initialize_all_variables.

suiyuan2009 replied:
good advice, I'll modify like this.

suiyuan2009 replied:
`var0 = tf.Variable(tf.zero_initializer(shape, dtype))` will double memory cost, which is what this pr wants to solve.

alextp replied:
It doesn't have to if zero_initializer is the op you wrote; the op passed to tf.Variable is just added to the initializers graph collection so it is run by tf.initialize_all_variables(). If it doesn't create any additional storage (and the op you wrote doesn't) then you should be fine. 

I also think we should test the memory usage by turning tracing on in the RunOptions and examining the step_stats.dev_stats.node_stats.allocator_used_memory.peak_bytes. That way the unit test will protect the code from changes which break this important property.

suiyuan2009 replied:
you can read mrry's comment, because tf.Variable will assign init_var to var, and the two tensors will stay in memory.

发自我的 iPhone

> 在 2016年9月13日，上午12:32，Alexandre Passos <notifications@github.com> 写道：
> 
> In tensorflow/contrib/framework/python/ops/variables_test.py:
> 
> > @@ -810,5 +810,27 @@ def testInitFromCheckpointWithScopes(self):
> >        self.assertAllEqual(init_value0, var0.eval())
> >        self.assertAllEqual(init_value1, var1.eval())
> >  
> > +class ZeroInitializerOpTest(tf.test.TestCase):
> > +
> > +  def _testZeroInitializer(self, shape, dtype):
> > +    var0 = tf.Variable(tf.zeros(shape, dtype=dtype))
> > +    var1 = tf.Variable(tf.ones(shape, dtype=dtype))
> > +    var0_zero = tf.contrib.framework.zero_initializer(var0)
> > +    var1_zero = tf.contrib.framework.zero_initializer(var1)
> It doesn't have to if zero_initializer is the op you wrote; the op passed to tf.Variable is just added to the initializers graph collection so it is run by tf.initialize_all_variables(). If it doesn't create any additional storage (and the op you wrote doesn't) then you should be fine.
> 
> I also think we should test the memory usage by turning tracing on in the RunOptions and examining the step_stats.dev_stats.node_stats.allocator_used_memory.peak_bytes. That way the unit test will protect the code from changes which break this important property.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

alextp replied:
So how about we add the unit test I suggested, checking memory usage, and then see whether we can make that syntax work?

suiyuan2009 replied:
yes, I'll add memory test.

发自我的 iPhone

> 在 2016年9月13日，上午1:16，Alexandre Passos <notifications@github.com> 写道：
> 
> In tensorflow/contrib/framework/python/ops/variables_test.py:
> 
> > @@ -810,5 +810,27 @@ def testInitFromCheckpointWithScopes(self):
> >        self.assertAllEqual(init_value0, var0.eval())
> >        self.assertAllEqual(init_value1, var1.eval())
> >  
> > +class ZeroInitializerOpTest(tf.test.TestCase):
> > +
> > +  def _testZeroInitializer(self, shape, dtype):
> > +    var0 = tf.Variable(tf.zeros(shape, dtype=dtype))
> > +    var1 = tf.Variable(tf.ones(shape, dtype=dtype))
> > +    var0_zero = tf.contrib.framework.zero_initializer(var0)
> > +    var1_zero = tf.contrib.framework.zero_initializer(var1)
> So how about we add the unit test I suggested, checking memory usage, and then see whether we can make that syntax work?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

alextp commented on tensorflow/contrib/framework/python/ops/variables_test.py:
```diff
@@ -810,5 +810,27 @@ def testInitFromCheckpointWithScopes(self):
+      self.assertAllClose(np.zeros(shape), var1.eval())
+
+  def testZeroInitializer(self):
+    for shape in ([10, 20], [10], [5,], [10, 20, 30]):
```
Why multiple shapes?

suiyuan2009 replied:
I'll modify this.

alextp commented on tensorflow/core/kernels/zero_initializer_op.h:
```diff
@@ -0,0 +1,70 @@
+class ZeroInitializerOp : public OpKernel {
+  public:
+    explicit ZeroInitializerOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("use_locking", &use_exclusive_lock_));
```
Why use_locking in a initializer? I think only one of these should run at a time anyway. I don't see the advantage of serializing them. Instead we should make this op fail if the variable is already initialized.

suiyuan2009 replied:
I thought `ctx->mutable_input` and `ctx->replace_ref_input` should be serialized, maybe I misunderstand `use_locking`.

alextp replied:
I was asking what is the value of use_locking=false here. In AssignOp use_locking=false lets more than one concurrent assign happen to the variable (which isn't helpful for assign but is for assign_add and even more so scatter_add where you're unlikely to see race conditions). There is no point in initializing the variable more than once concurrently, or assigning it while the initializer is running, so you should always grab the lock and not take use_locking as an attr.

suiyuan2009 replied:
I got it, thank you!

alextp commented on tensorflow/core/kernels/zero_initializer_op.h:
```diff
@@ -0,0 +1,70 @@
+    explicit ZeroInitializerOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("use_locking", &use_exclusive_lock_));
+      OP_REQUIRES(ctx, IsRefType(ctx->input_type(0)),
+          errors::InvalidArgument("input need to be a ref type"));
```
s/need/needs

alextp commented on tensorflow/core/kernels/zero_initializer_op.h:
```diff
@@ -0,0 +1,70 @@
+
+    void DoCompute(OpKernelContext* ctx) {
+      Tensor input = ctx->mutable_input(0, use_exclusive_lock_);
+      if (!input.IsInitialized()) {
```
op_requires here that the input is not initialized.

suiyuan2009 replied:
I'll modify it.

suiyuan2009 committed:
move code to tf.contrib

alextp commented on tensorflow/contrib/framework/kernels/zero_initializer_op.cc:
```diff
@@ -0,0 +1,37 @@
+      .TypeConstraint<T>("T"), \
+      ZeroInitializerOp<T>);
+#define REGISTER_CPU_KERNELS(T) REGISTER_KERNELS(CPU, T);
+TF_CALL_REAL_NUMBER_TYPES(REGISTER_CPU_KERNELS);
```
I think you can use TF_CALL_NUMBER_TYPES here

alextp replied:
Please ignore this comment.

alextp commented on tensorflow/contrib/framework/ops/variable_ops.cc:
```diff
@@ -0,0 +1,43 @@
+        return Status::OK();
+    })
+    .Doc(R"doc(
+Initialize 'ref' with all zeros. If 'ref' tensor is not initialized, the tensor
```
Update to say that the op requires that the tensor is not initialized.

alextp replied:
I think this doc change is the last thing this needs to be merged.

suiyuan2009 commented:
@alextp I moved all codes to tf.contrib.framework, and modified some code as you advised. I don't find a way to track memory usage of tensorflow in python, maybe the allocator in C++ can do it.

alextp commented:
Jenkins, test this please.

alextp commented on tensorflow/contrib/BUILD:
```diff
@@ -49,6 +49,7 @@ cc_library(
         "//tensorflow/contrib/layers:sparse_feature_cross_op_kernel",
         "//tensorflow/contrib/linear_optimizer:sdca_op_kernels",
         "//tensorflow/contrib/metrics:set_ops_kernels",
+        "//tensorflow/contrib/framework/kernels:all_kernels",
```
Sort alphabetically or the tests fail

suiyuan2009 committed:
modify doc and build file

alextp commented:
Jenkins, test this please.

alextp commented:
Can you rebase this with master? I see two conflicts, on contrib/.../variables.py and contrib/.../variables_test.py

alextp commented on tensorflow/contrib/BUILD:
```diff
@@ -61,6 +62,7 @@ cc_library(
         "//tensorflow/contrib/layers:sparse_feature_cross_op_op_lib",
         "//tensorflow/contrib/linear_optimizer:sdca_ops_op_lib",
         "//tensorflow/contrib/metrics:set_ops_op_lib",
+        "//tensorflow/contrib/framework:all_ops",
```
Sort alphabetically

suiyuan2009 commented:
@alextp conflict solved.

suiyuan2009 committed:
modify build file

alextp commented:
Jenkins, please test this change.

alextp commented:
Jenkins, test this please.

alextp commented on tensorflow/contrib/framework/BUILD:
```diff
@@ -24,7 +27,42 @@ py_library(
         "python/ops/variables.py",
     ],
+    data = [
+        ":python/ops/_variable_ops.so",
```
I think this dependency is not allowed, which is why a test fails (search for "fail" in https://ci.tensorflow.org/job/tensorflow-pull-requests-gpu/1833/consoleFull ).

Can you remove this dependency?

suiyuan2009 replied:
I find if `data` is removed, `tensorflow/contrib/framework/python/ops/variables_test.py` will failed with `ImportError: cannot import name gen_variable_ops`

alextp replied:
Weird. Then something else makes this test fail. Can you run it locally and debug?

suiyuan2009 replied:
Yes, I'll build a gpu version on linux.

alextp replied:
I don't think you need a GPU version, just bazel test on the entire directory should work (something like bazel test tensorflow/contrib/...)

suiyuan2009 replied:
weird, some rule is changed maybe. I run `bazel build -c opt --config=cuda //tensorflow/contrib/factorization:clustering_ops_test` and `bazel build -c opt --config=cuda //tensorflow/contrib/framework:ops_test`, they both failed.

alextp replied:
Can you try rebasing with master? Hopefully these were transient errors and are fixed now :-/

suiyuan2009 replied:
rebase can't help, I'm still trying, maybe `git revert` is the problem, I'll doing a `cherry-pick`.

suiyuan2009 replied:
`git revert` is not the problem... first I did `git fetch upstream master`, then `git rebase upstream/master` on master branch, then `git rebase origin/master` on a new branch which I cherry-pick some commits from this pr branch. test still failed.

alextp commented:
Jenkins, test this please.

alextp commented:
@gunan do you understand these test failures? Why is a deptest only failing on linux gpu? Is Jenkins building from a merge of master and this branch or from the head of this branch? I'd like to merge this PR but I don't know how to make it work.

martinwicke commented:
@keveman can you help with the build issue?

gunan commented:
ERROR: /workspace/tensorflow/contrib/framework/BUILD:38:1: in check_deps rule //tensorflow/contrib/framework:python/ops/_variable_ops.so_check_deps: 
Traceback (most recent call last):
	File "/workspace/tensorflow/contrib/framework/BUILD", line 38
		check_deps(name = 'python/ops/_variable_ops.so_check_deps')
	File "/workspace/tensorflow/tensorflow.bzl", line 669, in _check_deps_impl
		fail(_dep_label(input_dep) + " cannot...))
tensorflow/contrib/framework/kernels:zero_initializer_op cannot depend on tensorflow/core:framework.

here is the actual failure message.
Does this help about the issue any better?

alextp commented:
@gunan I did find this error message, but I don't understand why only the GPU build is doing this dependency check, and where was it added. Any clues?

suiyuan2009 commented:
@alextp I find `def tf_custom_op_library` has this deps check, and `//tensorflow/core:framework` in its `disallowed_deps` list. I use `tf_custom_op_library` rule in `tensorflow/contrib/framework/BUILD`, but my `tensorflow/contrib/framework/kernels:zero_initializer_op` is not depend on `tensorflow/core:framework` but `tensorflow/core:framework_headers_lib`.

suiyuan2009 committed:
a quick fix for deps check fail

suiyuan2009 commented:
@alextp works now.

alextp commented:
Jenkins, test this please.

alextp commented on tensorflow/contrib/framework/BUILD:
```diff
@@ -23,7 +26,41 @@ py_library(
+tf_custom_op_library(
+    name = "python/ops/_variable_ops.so",
+    srcs = [
+        "ops/variable_ops.cc",
```
Sort alphabetically (this has to be below the others) to get buildifier to stop complaining.

suiyuan2009 committed:
sort BUILD file labels

suiyuan2009 commented:
@alextp sorted now.

alextp commented:
Jenkins test this please.

On Sep 22, 2016 12:15, "Ziming Dong" <notifications@github.com> wrote:

> @alextp <https://github.com/alextp> sorted now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4077#issuecomment-248873534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxS0_Z0XR-UgE2reE0R9pDaReKEkfks5qsmNCgaJpZM4JuwXP>
> .
>

alextp commented:
Jenkins, test this please.

alextp commented:
@suiyuan2009 can you merge with master? It seem like this picked a version which fails some unrelated tests. I need the tests to pass before I can merge this PR.

suiyuan2009 commented:
@alextp I did a `git rebase`, so this is supposed to be the latest.

suiyuan2009 commented:
@alextp I configured python path to python3 and run bazel test the-target, everything is ok.

alextp commented:
Jenkins, test this please.

alextp commented:
Looking at the jenkins status I think tensorflow master might be broken. @m3bm3b as the buildcop, am I right or did I misread something?

alextp commented:
(and I suspect things are broken because https://ci.tensorflow.org/view/Tensorflow%20Jenkins%20Monitored%20builds/ has red in suspicious places)

m3bm3b commented:
It certainly looks suspicious to me.

I will look into it..

Mike



On 22 September 2016 at 10:43, Alexandre Passos <notifications@github.com>
wrote:

> (and I suspect things are broken because https://ci.tensorflow.org/
> view/Tensorflow%20Jenkins%20Monitored%20builds/ has red in suspicious
> places)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4077#issuecomment-248974836>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJsp2TFa3LuLGvtfIihowQIx2bU4HaHtks5qsr5XgaJpZM4JuwXP>
> .
>

gunan commented:
Opensource tests are triaged by another rotation.
Adding @martinwicke FYI.

alextp commented:
Jenkins, test this please.

suiyuan2009 commented:
@alextp can you see the build log, I used another machine to build and it failed, the log says `Error in `/usr/bin/python': double free or corruption`

suiyuan2009 commented:
test passed on my master branch, but failed on this pr branch, error msg is
`/tensorflow/bazel-out/local-py3-opt/testlogs/tensorflow/contrib/learn/random_forest_test/test.log
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmprlp8hgwp
WARNING:tensorflow:Using default config.
WARNING:tensorflow:Given features: Tensor("input:0", shape=(?, 4), dtype=float32), required signatures: TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None), Dimension(4)]), is_sparse=False).
WARNING:tensorflow:Given targets: Tensor("output:0", shape=(?,), dtype=float32), required signatures: TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None)]), is_sparse=False).
/home/zido/.cache/bazel/_bazel_zido/374c60a9747170d275c203f96a981f83/execroot/tensorflow/bazel-out/local-py3-opt/bin/tensorflow/contrib/learn/random_forest_test.runfiles/org_tensorflow/tensorflow/contrib/learn/python/learn/graph_actions.py:778: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
  logging.warn('Input iterator is exhausted: %s.', e)
WARNING:tensorflow:Input iterator is exhausted: .
.WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmp5qucovm4
WARNING:tensorflow:Using default config.
WARNING:tensorflow:Given features: Tensor("input:0", shape=(?, 13), dtype=float32), required signatures: TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None), Dimension(13)]), is_sparse=False).
WARNING:tensorflow:Given targets: Tensor("output:0", shape=(?,), dtype=float32), required signatures: TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None)]), is_sparse=False).
WARNING:tensorflow:Input iterator is exhausted: .
..
----------------------------------------------------------------------
Ran 3 tests in 17.033s

OK
*** Error in `/usr/bin/python3': double free or corruption (!prev): 0x000000000171c350 ***`

suiyuan2009 commented:
I run ` gdb --args python3 tensorflow/contrib/quantization/python/dequantize_op_test.py`, get
 ```
----------------------------------------------------------------------
Ran 3 tests in 0.070s

OK

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff787601f in _int_free (av=0x7ffff7bb5760 <main_arena>, p=<optimized out>, have_lock=0) at malloc.c:3996
3996    malloc.c: No such file or directory.
```

suiyuan2009 committed:
fix test failed

suiyuan2009 commented:
@alextp I think it's ok now. It seems `load_op_library` has some problem with `test_session`, but I can't figure it out.

alextp commented:
Jenkins, test this please.

suiyuan2009 commented:
`Android Demo App` test has a download problem.

alextp commented:
Jenkins, test this please.

gunan commented on tensorflow/contrib/framework/python/ops/variables_test.py:
```diff
@@ -1053,5 +1053,28 @@ def testMissingVariablesDict(self):
+        self.assertAllClose(np.zeros(shape), var.eval())
+
+  def testZeroInitializer(self):
+    for dtype in (tf.int32, tf.int64, tf.float32, tf.float64):
```
This test fails for tf.float32 and tf.float64 on GPU.
Is it supposed to not run on GPU, or is it a genuine bug?

suiyuan2009 replied:
It supports GPU,I've tested, but I've not tested on the current code.

suiyuan2009 replied:
What's the error?

gunan commented:
I do not see any error, It simply fails when I configure for GPU, and run:
$ bazel test tensorflow/contrib/framework:variables_test -c opt
--config=cuda

On Sat, Oct 1, 2016 at 9:05 PM, Ziming Dong <notifications@github.com>
wrote:

> *@suiyuan2009* commented on this pull request.
> ------------------------------
>
> In tensorflow/contrib/framework/python/ops/variables_test.py
> <https://github.com/tensorflow/tensorflow/pull/4077>:
>
> > +    var = tf.Variable(initializer)
> +    var_zero = tf.contrib.framework.zero_initializer(var)
> +    with self.test_session() as sess:
> +      with self.assertRaisesOpError("Attempting to use uninitialized value"):
> +        var.eval()
> +      if use_init:
> +        sess.run(var.initializer)
> +        with self.assertRaisesOpError("input is already initialized"):
> +          var_zero.eval()
> +        self.assertAllClose(np.ones(shape), var.eval())
> +      else:
> +        var_zero.eval()
> +        self.assertAllClose(np.zeros(shape), var.eval())
> +
> +  def testZeroInitializer(self):
> +    for dtype in (tf.int32, tf.int64, tf.float32, tf.float64):
>
> What's the error?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4077>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHlCOdln7PFHFtuDvcKzuodHX5G9Ciwtks5qvy1ygaJpZM4JuwXP>
> .
>

gunan commented:
Any nightly GPU build which runs this test is also broken:
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-mac-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=mac1-slave/79/console
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-mac-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=mac1-slave/79/console
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=gpu-linux/247/console
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3.5,label=gpu-linux/247/console
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=gpu-linux/247/console

it is not even flaky, it only runs on pip tests, and it fails on every
single run on GPU.
This test is omitted from no-pip runs on jenkins. That is how we missed the
breakage before.

On Sat, Oct 1, 2016 at 9:30 PM, Gunhan Gulsoy <gunan@google.com> wrote:

> I do not see any error, It simply fails when I configure for GPU, and run:
> $ bazel test tensorflow/contrib/framework:variables_test -c opt
> --config=cuda
>
> On Sat, Oct 1, 2016 at 9:05 PM, Ziming Dong <notifications@github.com>
> wrote:
>
>> *@suiyuan2009* commented on this pull request.
>> ------------------------------
>>
>> In tensorflow/contrib/framework/python/ops/variables_test.py
>> <https://github.com/tensorflow/tensorflow/pull/4077>:
>>
>> > +    var = tf.Variable(initializer)
>> +    var_zero = tf.contrib.framework.zero_initializer(var)
>> +    with self.test_session() as sess:
>> +      with self.assertRaisesOpError("Attempting to use uninitialized value"):
>> +        var.eval()
>> +      if use_init:
>> +        sess.run(var.initializer)
>> +        with self.assertRaisesOpError("input is already initialized"):
>> +          var_zero.eval()
>> +        self.assertAllClose(np.ones(shape), var.eval())
>> +      else:
>> +        var_zero.eval()
>> +        self.assertAllClose(np.zeros(shape), var.eval())
>> +
>> +  def testZeroInitializer(self):
>> +    for dtype in (tf.int32, tf.int64, tf.float32, tf.float64):
>>
>> What's the error?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tensorflow/tensorflow/pull/4077>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AHlCOdln7PFHFtuDvcKzuodHX5G9Ciwtks5qvy1ygaJpZM4JuwXP>
>> .
>>
>
>

suiyuan2009 commented:
@gunan I'll test on my local machine to see what's wrong.

jhseu commented:
Sent out a proposed fix for it internally (did it there because it's easier to test). Don't worry about it :)

Essentially just passed the Device as a parameter and changed it to:
`auto out_flat = out_tensor->flat<T>();`
`out_flat.device(ctx->eigen_device<Device>()) = out_flat.constant(T(0));`

suiyuan2009 commented:
@jhseu Thank you~
